adal/src/main/java/com/microsoft/aad/adal/AcquireTokenRequest.java

```diff
@@ -338,9 +340,29 @@ private AuthenticationResult acquireTokenSilentFlow(final AuthenticationRequest
         }
 
         // If we cannot switch to broker, return the result from local flow.
-        if (mBrokerProxy.canSwitchToBroker(authenticationRequest.getAuthority()) == BrokerProxy.SwitchToBroker.CANNOT_SWITCH_TO_BROKER
+        final BrokerProxy.SwitchToBroker switchToBrokerFlag = mBrokerProxy.canSwitchToBroker(authenticationRequest.getAuthority());
+        if (switchToBrokerFlag == BrokerProxy.SwitchToBroker.CANNOT_SWITCH_TO_BROKER
                 || !mBrokerProxy.verifyUser(authenticationRequest.getLoginHint(), authenticationRequest.getUserId())) {
             return authResult;
+        } else if (switchToBrokerFlag == BrokerProxy.SwitchToBroker.NEED_PERMISSIONS_TO_SWITCH_TO_BROKER) {
+            //For android M and above
+            throw new UsageAuthenticationException(
+                        ADALError.DEVELOPER_BROKER_PERMISSIONS_MISSING,
+                        "Broker related permissions are missing for GET_ACCOUNTS");
+        }
+
+        //check if App gives permissions of broker usage
```

Comment:
BrokerProxy.canSwitchToBroker() only checks the GET_ACCOUNTS permission for Android 23+ which need to be granted when app is running. But we also need to check permissions for Android 22 and below, which are set in the manifest. We did not check the permissions required for both Android 23+ ans Android 22-. My current concern is if we should check permissions for Android 22- at the very beginning of the acquireTokenSilent call, because these permissions need to be set by the developer in the manifest if using broker.